JENA-1267: RDFConnection

afs opened the pull request:
An API that brings together SPARQL APIs.

2 new modules:
- jena-rdfconnection
- jena-integration-tests

afs commented:
The "test" module is needed because the tests need a Fuseki test server but  jena-rdfconnection is before Fuseki in the build sequence.

Some tests in jena-fuseki-core that are DatasetAccess and query, update SPARQL API calls could similarly be moved to this new location.

afs commented:
Documentation: https://github.com/afs/jena/blob/rdf-connection/jena-rdfconnection/Documentation.md

ajs6f commented on pom.xml:
```diff
@@ -155,6 +159,12 @@
         <module>jena-fuseki2</module>
         <module>jena-permissions</module>
 
+        <!-- Tests of artifacts that require 
```
Is this comment missing some text?

ajs6f commented:
Is the basic idea of `jena-integration-tests` to provide a place for inter-module tests?

rvesse commented on jena-integration-tests/README.md:
```diff
@@ -0,0 +1,11 @@
+Jena Intrgration Tests
```
Typo r -> e

rvesse commented on jena-integration-tests/src/test/java/org/apache/jena/test/rdfconnection/AbstractTestRDFConnection.java:
```diff
@@ -0,0 +1,382 @@
+    static String DIR = "testing/RDFConnection/" ;
+    
+    protected abstract RDFConnection connection() ;
+    // Not all RDFConenction types support.
```
Typo: conenction -> connection

rvesse commented on jena-integration-tests/src/test/java/org/apache/jena/test/rdfconnection/AbstractTestRDFConnection.java:
```diff
@@ -0,0 +1,382 @@
+    
+    protected abstract RDFConnection connection() ;
+    // Not all RDFConenction types support.
+    //  (may acquite RDFConnection.supportTransactionalAbort but ATM that isn't included)
```
Typo: acquite -> acquire

rvesse commented on jena-rdfconnection/Documentation.md:
```diff
@@ -0,0 +1,205 @@
+responsibility of the remote server.
+
+An `RDFConenction` will at least provide the client-side locking features.
+This means that overlapping operations that chnage data are naturally
```
Typo: chnage -> change

rvesse commented on jena-rdfconnection/Documentation.md:
```diff
@@ -0,0 +1,205 @@
+When creating an `QueryExecution` explicitly, care shoud be taken to close
+it. If the application wishes to capture the result set from a SELECT query and
+retain it across the lifetime of the transaction or `QueryExecution`, then
+the application create a copy which is not attached to any external system
```
Missing word: the application *should* create

rvesse commented on jena-rdfconnection/Documentation.md:
```diff
@@ -0,0 +1,205 @@
+
+In addition to the SPARQL Graph Store Protocol, operations on whole
+datasets are provided for fetching (HTTP GET), adding data (HTTP POST) and
+settign the data (HTTP PUT) on a dataset URL.  This assumes the remote
```
Typo: settign -> setting

rvesse commented on jena-rdfconnection/Documentation.md:
```diff
@@ -0,0 +1,205 @@
+
+## Examples
+
+https://github.com/afs/jena-rdfconnection/tree/master/src/main/java/rdfconnection/examples
```
TODO: update this to official repository after merging

afs replied:
Done, to avoid forgetting later.

rvesse commented on jena-rdfconnection/src/main/java/org/apache/jena/rdfconnection/RDFConnectionFactory.java:
```diff
@@ -0,0 +1,84 @@
+    /** Create a connection to a remote location by URL.
+     * This is the URL for the dataset.
+     * 
+     *  This call requires specifying the names of the servicerelative to the dataset URL.
```
Missing space

rvesse commented on jena-rdfconnection/src/main/java/org/apache/jena/rdfconnection/RDFConnectionRemote.java:
```diff
@@ -0,0 +1,478 @@
+        this.svcQuery = formServiceURL(destination,sQuery) ;
+        this.svcUpdate = formServiceURL(destination,sUpdate) ;
+        this.svcGraphStore = formServiceURL(destination,sGSP) ;
+//        if ( httpClient == null )
```
I am guessing that the null check is unnecessary because passing a null client to `HttpOp` causes the default client to be used?

afs commented:
Thank you for the reviews - types corrected.

afs commented:
> Is the basic idea of jena-integration-tests to provide a place for inter-module tests?

The driver is that testing remote operations needs a SPARQL server but that is not a dependency of the code, only of the tests.  However, Fuseki isn't built at this point in the build process.

jena-rdfconnection goes into the apache-jena-libs and apache-jena artifacts but Fuseki is built after those.

This must be a common need in system so if there is a better common pattern, then I'm happy to switch to that.

There are a number of ways this could be done:

1. Build modules in a different order - e.g. do building steps before delivery modules.
2. Put the tests in Fuseki
3. As here, a specific module.
4. Use a previous version of Fuseki.

(1) works in this case but it may not work out whereas the idea of an integration testing module, seems general. (2) is a less clean version of (1); this is already done in a small way - those can be moved now. (4) seems accident prone.

So (3) looks the best way and may be applicable for other situations. It may even to help unblock evolution of the legacy module/packing structure (the old Turtle reader in jena-core is needed for assembler tests - I haven't had a chance to try yet).

afs commented:
I wasn't happy with a single fixed way to use a local connection - I think all of copy (pseudo remote), no isolation and read-only (no copy so a pragmatic balance for larger datasets) make sense in different situations.

so I have added two isolation modes added to the documentation and added tests.

ajs6f commented:
I am completely cool with the idea of separate i-test modules-- I think that's very much the right idea. Is the thinking that we can move it later in the build as need be (i.e. if we wanted to include some i-tests for some later modules)?

afs commented:
Its location in the build shouldn't matter too much, just after all the modules it is testing.

ajs6f commented:
Right, cool. Just wanted to make sure we leave the door open to move it later in the build to include testing of later modules. +1
